storage: expose experimental clockless reads mode

tbg opened the pull request:
See #14093. Add the `--max-offset=experimental-clockless` mode in which the
cluster sends an empty request through Raft for each read and always restarts
on uncertainty errors.

Not for production use.
Not tested (#15738 will help).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16867)
<!-- Reviewable:end -->

bdarnell commented:
Looks good, but I'd rather get rid of the explicit mentions of MaxInt64 scattered everywhere.

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cli/cliflags/flags.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJuX1gHORNehhj7yk2:-KoJuX1gHORNehhj7yk3:b4bdqo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/cli/cliflags/flags.go#L342)):*
> ```Go
> 	//
> 	// </PRE>
> 	// Specifying the string value 'clockless' instead of a duration runs the cluster in
> ```

s/clockless/experimental-clockless/

---

*[pkg/internal/client/lease.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJunB5dOpI03kPYjpZ:-KoJunB5dOpI03kPYjp_:brlfaei) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/internal/client/lease.go#L126)):*
> ```Go
> func (m *LeaseManager) timeRemaining(val *LeaseVal) time.Duration {
> 	maxOffset := m.clock.MaxOffset()
> 	if maxOffset == time.Duration(math.MaxInt64) {
> ```

Make a constant for `time.Duration(math.MaxInt64)`. Or even better, add an `IsClockless() bool` method to `hlc.Clock`

---

*[pkg/kv/txn_coord_sender.go, line 472 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJv_bQ6UIW4Wvahiug:-KoJv_bQ6UIW4Wvahiuh:b-1kfgr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/kv/txn_coord_sender.go#L472)):*
> ```Go
> 	// likely the first one unless a transaction commits with an odd timestamp.
> 	//
> 	// Can't use the linearizable optimization with clockless reads since in
> ```

`--linearizable` is not really an optimization; it's a "hack" or a "technique" depending on how charitable you're feeling. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-:-KoJykWbeiY3lTxcSeqv:b3i30ww)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The change is remarkably straightforward. I wonder if this should be called "quorum read mode".

bdarnell commented:
We have an [ancient and unimplemented placeholder for a `CONSENSUS` read mode](https://github.com/cockroachdb/cockroach/blob/99f445102cf6672239a96a752bb9d87be6c19850/pkg/roachpb/api.proto#L37-L41) that can be applied on a per-read basis. That's not quite the same as this fully-clockless mode. 

Would it ever make sense to offer quorum/consensus reads on a per-operation basis in our normal mode? If so, it might be better to reserve the term "quorum reads" for that mode, and continue calling this "clockless". If we don't think we'd want to offer this as a separate feature, then I'd be fine with calling this "quorum reads" when it graduates from experimental status.

petermattis commented:
Wouldn't the `CONSENSUS` read mode do something similar to this PR? If yes, `--max-offset=consensus` (or whatever we call it) could simply be seen as changing the default mode for reads.

bdarnell commented:
`CONSENSUS` reads as envisioned by that option would do part of this PR (the part of replica.go to always go through raft), but not all of it (e.g. the change in data.go to alter the handling of clock uncertainty-related transaction restarts)

tschottdorf commented:
Seeing that it's experimental, I left the naming as is.
PTAL

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cli/cliflags/flags.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJuX1gHORNehhj7yk2:-KoK-OaWSyBkV532mUe-:b-pwsc0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/cli/cliflags/flags.go#L342)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/clockless/experimental-clockless/
</blockquote></details>

By the time this comment gets put in, it'll be "clockless" (or something else altogether), no?

---

*[pkg/internal/client/lease.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJunB5dOpI03kPYjpZ:-KoOndaQFIEs1dvivGMp:bkrjizx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/internal/client/lease.go#L126)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Make a constant for `time.Duration(math.MaxInt64)`. Or even better, add an `IsClockless() bool` method to `hlc.Clock`
</blockquote></details>

Seeing that it's already awkward that `hlc.Clock` knows anything about `MaxOffset`, I went with the constant.

---

*[pkg/kv/txn_coord_sender.go, line 472 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJv_bQ6UIW4Wvahiug:-KoOnalcP_16yWxuHyBt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/kv/txn_coord_sender.go#L472)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`--linearizable` is not really an optimization; it's a "hack" or a "technique" depending on how charitable you're feeling. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-:-KoOqo2fUYG5Hy23xhWM:bxhquzo)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 17 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/cliflags/flags.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJuX1gHORNehhj7yk2:-KoOwaU0o0Kp9aVIZojV:b-vtp7mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/cli/cliflags/flags.go#L342)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

By the time this comment gets put in, it'll be "clockless" (or something else altogether), no?
</blockquote></details>

Before it's in the help text, yes, but this comment is here now. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-:-KoOwtcBdGL9sGASMCOw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: expose experimental clockless reads mode

See #14093. Add the `--max-offset=experimental-clockless` mode in which the
cluster sends an empty request through Raft for each reads and always restarts
on uncertainty errors.

Not for production use.
Not tested (#15738 will help).

tschottdorf commented:
TFTRs!

---

Review status: 3 of 17 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cli/cliflags/flags.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-KoJuX1gHORNehhj7yk2:-KoOyamhY_aghr2Iti6z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d6ae4bf811484dca16865902b1431ea4fa40da9/pkg/cli/cliflags/flags.go#L342)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Before it's in the help text, yes, but this comment is here now. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16867#-:-KoOythrKi0LJqDcnm81:bz5mair)*
<!-- Sent from Reviewable.io -->
